src/REP15.sol

```diff
@@ -0,0 +1,466 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import { IERC165, ERC165 } from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import { IERC721, ERC721 } from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import { IREP15 } from "./interfaces/IREP15.sol";
+import { IREP15Errors } from "./interfaces/IREP15Errors.sol";
+import { IREP15ContextCallback } from "./interfaces/IREP15ContextCallback.sol";
+import { REP15Utils } from "./REP15Utils.sol";
+
+abstract contract REP15 is ERC721, IREP15, IREP15Errors {
+  using REP15Utils for REP15Utils.Delegation;
+  using REP15Utils for REP15Utils.Context;
+  using REP15Utils for REP15Utils.TokenContext;
+
+  uint64 internal immutable _maxDetachingDuration;
+
+  mapping(uint256 tokenId => REP15Utils.Delegation) private _delegations;
```

Comment:
```suggestion
  mapping(uint256 tokenId => REP15Utils.Delegation) private _delegation;
```

Reply:
I see many programming languages use plural nouns for mapping.

Reply:
might cause confusing with array types imo

Reply:
It is not important, we can still understand the logic. imo